be/src/exec/pipeline/query_context.cpp

```diff
@@ -34,21 +35,60 @@ void QueryContext::cancel(const Status& status) {
     _fragment_mgr->cancel(status);
 }
 
-static constexpr size_t QUERY_CONTEXT_MAP_SLOT_NUM = 64;
-static constexpr size_t QUERY_CONTEXT_MAP_SLOT_NUM_MASK = (1 << 6) - 1;
+QueryContextManager::QueryContextManager(size_t log2_num_slots)
+        : _num_slots(1 << log2_num_slots),
+          _slot_mask(_num_slots - 1),
+          _mutexes(_num_slots),
+          _context_maps(_num_slots),
+          _second_chance_maps(_num_slots) {}
+
+Status QueryContextManager::init() {
+    try {
+        _clean_thread = std::make_shared<std::thread>(_clean_func, this);
+        return Status::OK();
+    } catch (...) {
+        return Status::InternalError("Fail to create clean_thread of QueryContextManager");
+    }
+}
+
+void QueryContextManager::_clean_query_contexts() {
+    for (auto i = 0; i < _num_slots; ++i) {
+        auto& mutex = _mutexes[i];
+        auto& sc_map = _second_chance_maps[i];
+        std::unique_lock write_lock(mutex);
+        auto sc_it = sc_map.begin();
+        while (sc_it != sc_map.end()) {
+            if (sc_it->second->has_no_active_instances() && sc_it->second->is_expired()) {
+                sc_it = sc_map.erase(sc_it);
+            } else {
+                ++sc_it;
+            }
+        }
+    }
+}
+
+void QueryContextManager::_clean_func(QueryContextManager* manager) {
```

Comment:
Does this lazy cleanup mechanism cause cross-thread memory release issues?

Reply:
this thread only reap queries that finishes abnormally in the case of partial instances are delivered successfully while partial ones are not. so the number is rare.